cpp/src/core/library.cpp

```diff
@@ -69,6 +69,47 @@ class Mapper : public legate::mapping::Mapper {
     return mappings;
   }
 
+  std::optional<std::size_t> allocation_pool_size(const legate::mapping::Task& task,
+                                                  legate::mapping::StoreTarget memory_kind)
+  {
+    const auto task_id = static_cast<int>(task.task_id());
+
+    if (memory_kind == legate::mapping::StoreTarget::ZCMEM) {
+      switch (task_id) {
+        case legate::dataframe::task::OpCode::Join:
+          // TODO: Join is identical to GroupBy, but we would have to look at
+          // both input tables to ge the maximum column number.
+          return std::nullopt;
+        case legate::dataframe::task::OpCode::GroupByAggregation: {
+          // Join and Aggregation use repartitioning and use ZCMEM for NCCL.
+          // This depends on the number of columns (first argument to task
+          // is a vector of these columns).
+          // Join repartitions the table one after the other so one is enough.
+          if (task.is_single_task()) {
+            // No need for repartitioning, so no need for ZCMEM
+            return 0;
+          }
+          auto num_cols = task.scalars().at(0).value<int32_t>();
+          auto nrank    = task.get_launch_domain().get_volume();
+
+          // Space for the exchange buffers containing size chunks:
+          size_t size_exchange_nbytes = nrank * nrank * 2 * sizeof(std::size_t);
+          // Space for column packing metadata exchange
+          size_t sizeof_serialized_column = 6 * 8;  // not public, rough upper bound
+          // num_cols * 2 for the string column child and + 1 for the number of cols
+          size_t metadata_nbytes = nrank * (num_cols * 2 + 1) * sizeof_serialized_column;
+
+          return size_exchange_nbytes + metadata_nbytes;
```

Comment:
Not sure this is really worthwhile, since the initial PR had the `return 0` here, I thought I would get the correct value.

For aggregations that is nice, for `join` would have to fetch both tables, although that shouldn't be too bad (we just don't have a convenience function to do that via the task).

Locally, I see cuda allocation failures in some join tasks, it isn't clear to me yet why those happen, though.